src/main/kotlin/org/jitsi/nlj/MediaSourceDesc.kt

```diff
@@ -0,0 +1,161 @@
+ *
+ * @author George Politis
+ */
+class MediaSourceDesc
+@JvmOverloads constructor(
+    /**
+     * The [RtpEncodingDesc]s that this [MediaSourceDesc]
+     * possesses, ordered by their subjective quality from low to high.
+     */
+    val rtpEncodings: Array<RtpEncodingDesc>,
+    /**
+     * A string which identifies the owner of this source (e.g. the endpoint
+     * which is the sender of the source).
+     */
+    val owner: String? = null
+) {
+    /**
+     * Allow the lookup of a layer by the encoding id of a received packet.
+     */
+    private val layersById: MutableMap<Long, RtpLayerDesc> = HashMap()
+
+    /**
+     * Allow the lookup of a layer by index.
+     */
+    private val layersByIndex: MutableMap<Int, RtpLayerDesc> = HashMap()
+
+    /**
+     * Update the layer cache.  Should be synchronized on [this].
+     */
+    private fun updateLayerCache() {
+        layersById.clear()
+        layersByIndex.clear()
+
+        for (encoding in rtpEncodings) {
+            for (layer in encoding.layers) {
+                layersById[encoding.encodingId(layer)] = layer
+                layersByIndex[layer.index] = layer
+            }
+        }
+    }
+
+    init { updateLayerCache() }
+
+    /**
+     * Gets the last "stable" bitrate (in bps) of the encoding of the specified
+     * index. The "stable" bitrate is measured on every new frame and with a
+     * 5000ms window.
+     *
+     * @return the last "stable" bitrate (bps) of the encoding at the specified
+     * index.
+     */
+    fun getBitrateBps(nowMs: Long, idx: Int): Long {
+        val layer = getRtpLayerByQualityIdx(idx) ?: return 0
+
+        // TODO: previous code returned a lower layer if this layer's bitrate was 0.
+        // Do we still need this?
+        return layer.getBitrateBps(nowMs)
+    }
+
+    fun hasRtpLayers(): Boolean =
+        synchronized(this) {
+            !layersByIndex.isEmpty()
+        }
+
+    /**
+     * Get an iterator over the source's RTP layers, in quality order.  Should be synchronized on [this].
+     */
+    val rtpLayers: Iterable<RtpLayerDesc>
+        get() = Iterable { Arrays.stream(rtpEncodings).flatMap { e: RtpEncodingDesc -> Arrays.stream(e.layers) }.iterator() }
+
+    fun numRtpLayers(): Int =
+        synchronized(this) {
+            layersByIndex.size
+        }
+
+    val primarySSRC: Long
+        get() = rtpEncodings[0].primarySSRC
+
+    fun getRtpLayerByQualityIdx(idx: Int): RtpLayerDesc? =
+        synchronized(this) {
+            layersByIndex[idx]
+        }
+
+    fun findRtpLayerDesc(videoRtpPacket: VideoRtpPacket): RtpLayerDesc? {
+        synchronized(this) {
+            if (ArrayUtils.isNullOrEmpty(rtpEncodings)) {
+                return null
+            }
+            val encodingId = getEncodingId(videoRtpPacket)
+            val desc = layersById[encodingId]
+            if (desc != null) {
+                return desc
+            }
+            /* ??? Does this part actually get used? */
```

Comment:
Did you determine the answer to this?

Reply:
No; this is old code that I copied over, which I was uncertain about, thus the added comment.

Reply:
Updated the comment so it's clearer that I'm uncertain. 🙂